falcon/asgi/request.py

```diff
@@ -364,20 +368,25 @@ def __init__(self, scope, receive, first_event=None, options=None):
             # NOTE(kgriffs): According to ASGI 3.0, header names are always
             #   lowercased, and both name and value are byte strings. Although
             #   technically header names and values are restricted to US-ASCII
-            #   we decode using the default 'utf-8' because it is a little
-            #   faster than passing an encoding option.
-            header_name = header_name.decode()
-            header_value = header_value.decode()
+            #   we decode later (just-in-time) using the default 'utf-8' because
+            #   it is a little faster than passing an encoding option (except
+            #   under Cython).
+            #
+            #   The reason we wait to decode is that the typical app will not
+            #   need to decode all request headers, and we usually can just
+            #   leave the header name as a byte string and look it up that way.
+            #
 
             # NOTE(kgriffs): There are no standard request headers that
             #   allow multiple instances to appear in the request while also
             #   disallowing list syntax.
-            if header_name not in req_headers or header_name in SINGLETON_HEADERS:
+            if header_name not in req_headers or header_name in _SINGLETON_HEADERS_BYTESTR:
                 req_headers[header_name] = header_value
             else:
-                req_headers[header_name] += ',' + header_value
+                req_headers[header_name] += b',' + header_value
```

Comment:
bytes don't have the same optimizations as strings regarding avoiding creating a new object each time, so maybe we could see if append+join is faster? The downside would be that it's a bit ugly tracking which header needs joining and which does not.

Reply:
Good thought. Perhaps we should create an issue to look at this later? I suspect that most request headers for most apps won't have many (if any) duplicate header lines.

Reply:
I've added an issue with some potential additional points discussed here and on gitter wrt this issue
https://github.com/falconry/falcon/issues/1870